### PR TITLE
PARQUET-1404: [C++] Adding Page level Indexes

### PR DESCRIPTION
**Pull request by a2un**
Implementation as per the doc in Parquet-922
https://docs.google.com/document/d/1sBACp8Lbutuj1Zxdowvsrlm8ku4BFxf8U_Do5K2wSO4/edit#

### CODE REVIEW

File: cpp/src/parquet/metadata.h [hunk 1]
```diff
@@ -24,10 +24,15 @@
 #include <vector>
 
 #include "arrow/util/key_value_metadata.h"
-
+#include <boost/any.hpp>
```
Comment (wesm) [inline cpp/src/parquet/metadata.h:27]: Please don't add boost includes to any public headers

File: cpp/examples/parquet/low-level-api/run-imapala-queries.sh [hunk 2]
```diff
@@ -0,0 +1,122 @@
+source $IMPALA_HOME/bin/impala-config.sh
```
Comment (wesm) [inline cpp/examples/parquet/low-level-api/run-imapala-queries.sh:1]: I don't think this file should be here?

File: cpp/src/parquet/column_writer.h [hunk 3]
```diff
@@ -153,6 +155,11 @@ class TypedColumnWriter : public ColumnWriter {
   // column.
   virtual void WriteBatch(int64_t num_values, const int16_t* def_levels,
                           const int16_t* rep_levels, const T* values) = 0;
+  
+  // Write a batch of repetition levels, definition levels, and values to the
+  // column.
+  virtual void WriteBatchWithIndex(int64_t num_values, const int16_t* def_levels,
+                          const int16_t* rep_levels, const T* values) = 0;
```
Comment (wesm) [inline cpp/src/parquet/column_writer.h:162]: What does this API do differently from `WriteBatch`?

File: cpp/src/parquet/file_reader.h [hunk 4]
```diff
@@ -43,6 +43,7 @@ class PARQUET_EXPORT RowGroupReader {
   struct Contents {
     virtual ~Contents() {}
     virtual std::unique_ptr<PageReader> GetColumnPageReader(int i) = 0;
+    virtual std::unique_ptr<PageReader> GetColumnPageReaderWithIndex(int i,void* predicate, int64_t& min_index, int predicate_Col, int64_t& row_index,Type::type type_num) = 0;
```
Comment (wesm) [inline cpp/src/parquet/file_reader.h:46]: Add a comment explaining what this function does and what the arguments mean. The opaque `void* predicate` needs to be explained, for example

Note we don't use mutable references for arguments in this project

File: cpp/src/parquet/file_reader.h [hunk 5]
```diff
@@ -56,8 +57,12 @@ class PARQUET_EXPORT RowGroupReader {
   // column. Ownership is shared with the RowGroupReader.
   std::shared_ptr<ColumnReader> Column(int i);
 
+  std::shared_ptr<ColumnReader> ColumnWithIndex(int i,void* predicate, int64_t& min_index, int predicate_col, int64_t& row_index,Type::type type_num);
```
Comment (wesm) [inline cpp/src/parquet/file_reader.h:60]: What does this method do, and what do the arguments mean? What is `void* predicate`?

File: cpp/src/parquet/file_reader.h [hunk 6]
```diff
@@ -56,8 +57,12 @@ class PARQUET_EXPORT RowGroupReader {
...
+
   std::unique_ptr<PageReader> GetColumnPageReader(int i);
 
+  std::unique_ptr<PageReader> GetColumnPageReaderWithIndex(int column_index, void* predicate, int64_t& min_index , int predicate_col, int64_t& row_index,Type::type type_num);
```
Comment (wesm) [inline cpp/src/parquet/file_reader.h:64]: Same question

**Review by wesm — COMMENTED** [review]
I'm still having a hard time understanding the public APIs that are proposed. Let's focus for the moment on the header files and documenting how the new APIs work / what their arguments mean

File: cpp/examples/parquet/low-level-api/run-imapala-queries.sh (hunk 2 above)
Comment (a2un) [reply cpp/examples/parquet/low-level-api/run-imapala-queries.sh:1]: Yes, I will be removing after completing my tests

File: cpp/src/parquet/column_writer.h (hunk 3 above)
Comment (a2un) [reply cpp/src/parquet/column_writer.h:162]: I want to WriteBatch, but adding Column Index and Offset Index to the end of the of Parquet file, I will be updating the documentation

File: cpp/src/parquet/file_reader.h (hunk 4 above)
Comment (a2un) [reply cpp/src/parquet/file_reader.h:46]: I wanted to add this function to utilize a predicate in scanning the column indices for skipping to the page that could contain the value

File: cpp/src/parquet/file_reader.h (hunk 5 above)
Comment (a2un) [reply cpp/src/parquet/file_reader.h:60]: predicate is the value that gets pushed down from the query engine, to be used to compared with column indices, and detect the page that could potentially contain the value of the predicate, I used void* for support for different parquet data types

File: cpp/src/parquet/file_reader.h (hunk 6 above)
Comment (a2un) [reply cpp/src/parquet/file_reader.h:64]: predicate is the value that gets pushed down from the query engine, to be used to compared with column indices, and detect the page that could potentially contain the value of the predicate, I used void* for support for different parquet data types

File: cpp/src/parquet/file_reader.h (hunk 4 above)
Comment (a2un) [reply cpp/src/parquet/file_reader.h:46]: predicate is the value that gets pushed down from the query engine, to be used to compared with column indices, and detect the page that could potentially contain the value of the predicate, I used void* for support for different parquet data types

File: cpp/src/parquet/metadata.h (hunk 1 above)
Comment (a2un) [reply cpp/src/parquet/metadata.h:27]: I am not using boost, I think the final commit doesn't have that header

File: cpp/examples/parquet/low-level-api/run-imapala-queries.sh (hunk 2 above)
Comment (a2un) [reply cpp/examples/parquet/low-level-api/run-imapala-queries.sh:1]: it is not part of the file PR

**Comment by ggershinsky** [comment]
> github-actions bot commented 4 days ago
> https://issues.apache.org/jira/browse/PARQUET-1401

I'm curious why.

**Comment by ggershinsky** [comment]
This pull request seems to lack encryption of column indexes and column offsets. For an example of how it is done in Java, see https://github.com/apache/parquet-mr/pull/776

**Comment by wesm** [comment]
> I'm curious why. This sends me all comments etc.

This is to help contributors who want to navigate easily back to the JIRA issue

**Comment by ggershinsky** [comment]
> > I'm curious why. This sends me all comments etc.
> 
> This is to help contributors who want to navigate easily back to the JIRA issue

Looks like this pull requests is named after a wrong JIRA, https://issues.apache.org/jira/browse/PARQUET-1401 is about "RowGroup offset and total compressed size fields". Given the source branch, it should have been https://issues.apache.org/jira/browse/PARQUET-1404.

**Comment by wesm** [comment]
The GitHub action uses the GitHub PR title to match with the JIRA, I just fixed the title to be PARQUET-1404

**Comment by wesm** [comment]
@a2un it seems this patch is still very much WIP, would it be alright to close the PR and you can re-open when you have something closer to review/merge-ready for us to review? CI builds will still run on your fork

**Comment by a2un** [comment]
> @a2un it seems this patch is still very much WIP, would it be alright to close the PR and you can re-open when you have something closer to review/merge-ready for us to review? CI builds will still run on your fork

Okay, I just wanted a place to hold discussion and ask you questions on the project

**Comment by wesm** [comment]
We can discuss more on the mailing list -- now that I know where the branch is, if you want me to take another look at the headers per the comments above let me know

**Comment by malthe** [comment]
I added an issue https://issues.apache.org/jira/browse/ARROW-10158.

**Pull request closed without merging**